source/agora/test/QuorumShuffle.d

```diff
@@ -0,0 +1,469 @@
+    foreach (idx, node; nodes.enumerate)
+        writefln("Node %s: %s\n", idx, node.getQuorumConfig);
+
+    nodes.enumerate.each!((idx, node) =>
+        retryFor(node.getQuorumConfig() == quorums_1[idx], 5.seconds,
+            format("Node %s has quorum config %s. Expected: %s",
+                idx, node.getQuorumConfig(), quorums_1[idx])));
+
+    // check that the preimages were revealed before we trigger block creation
+    // note: this is a workaround. A block should not be accepted if there
+    // are no preimages for this height - however currently block signatures
+    // are not implemented yet (#365)
+    const enrollments = network.blocks[0].header.enrollments;
+    void checkDistance (uint distance)
+    {
+        nodes.each!(node =>
+            enrollments.each!(enr =>
+                retryFor(node.getPreimage(enr.utxo_key).distance >= distance,
+                    5.seconds)));
+    }
+
+    checkDistance(3);
+
+    auto txs = makeChainedTransactions(WK.Keys.Genesis,
+        network.blocks[$ - 1].txs, 1);
+    txs.each!(tx => nodes[0].putTransaction(tx));
+
+    // at block height 3 a shuffle should occur
+    nodes.enumerate.each!((idx, node) =>
+        retryFor(node.getBlockHeight() == 3, 5.seconds,
+            format("Node %s has block height %s. Expected: %s",
+                idx, node.getBlockHeight(), 3)));
+
+    enum quorums_2 = [
+        // 0
+        QuorumConfig(4, [
+            WK.Keys.D.address,
+            WK.Keys.A.address,
+            WK.Keys.C.address,
+            WK.Keys.E.address]),
+
+        // 1
+        QuorumConfig(4, [
+            WK.Keys.F.address,
+            WK.Keys.B.address,
+            WK.Keys.A.address,
+            WK.Keys.E.address]),
+
+        // 2
+        QuorumConfig(4, [
+            WK.Keys.D.address,
+            WK.Keys.F.address,
+            WK.Keys.B.address,
+            WK.Keys.C.address]),
+
+        // 3
+        QuorumConfig(4, [
+            WK.Keys.D.address,
+            WK.Keys.F.address,
+            WK.Keys.C.address,
+            WK.Keys.E.address]),
+
+        // 4
+        QuorumConfig(4, [
+            WK.Keys.B.address,
+            WK.Keys.A.address,
+            WK.Keys.C.address,
+            WK.Keys.E.address]),
+
+        // 5
+        QuorumConfig(4, [
+            WK.Keys.D.address,
+            WK.Keys.F.address,
+            WK.Keys.A.address,
+            WK.Keys.C.address]),
+    ];
+
+    version (none)
+    foreach (idx, node; nodes.enumerate)
+        writefln("Node %s: %s\n", idx, node.getQuorumConfig);
+
+    nodes.enumerate.each!((idx, node) =>
+        retryFor(node.getQuorumConfig() == quorums_2[idx], 5.seconds,
+            format("Node %s has quorum config %s. Expected: %s",
+                idx, node.getQuorumConfig(), quorums_2[idx])));
+
+    checkDistance(6);  // make sure we can create the next 3 blocks
+
+    void createBlock (ulong height)
+    {
+        txs = makeChainedTransactions(WK.Keys.Genesis, txs, 1);
```

Comment:
`TxBuilder` ?

Reply:
Sure. I'll fix it.

Reply:
👀

Reply:
Oh I just realized this is a duplicate function of `makeBlock` above. Heh.